### ISSUE TITLE
Running update with scriptWriter marks changesets as run

### ISSUE THREAD

**Issue by leblonk** [liquibase/liquibase#2579]
Calling
```
    Writer scriptWriter = new StringWriter();
    liquibase.update(new Contexts(), scriptWriter);
```
should just log the statements it would run to the scriptWriter, and it does that in v3.5.3. In 4.5.0&4.7.1 though, it not only logs the statements, but it also marks the any new changelogs as completed in the DATABASECHANGELOG table.
I have not tried any other versions to further narrow down when this broke.

This line
https://github.com/liquibase/liquibase/blob/79c0e38fe8987bdc7f4200cc05efa1b7c8276248/liquibase-core/src/main/java/liquibase/changelog/visitor/UpdateVisitor.java#L67
calls a method that uses a different executor (a live jdbc one) to update the DATABASECHANGELOG table, instead of using a LoggingExecutor

To replicate
- start with database (mysql if that matters) and changelog file (.sql if that matters) with at least one un-run changelog
- run:
```
    Writer scriptWriter = new StringWriter();
    liquibase.update(new Contexts(), scriptWriter);  // log only
    liquibase.update(new Contexts());  // do the actual update
```
- notice that the new changelogs are marked as run, but the actual changes have not been made in the database

**Comment by leblonk** [comment]
Added, this is apparently trigger by instantiating a Liquibase instance twice. There are some static caches (like LockServiceFactory) that will keep the first `database` instance, and others that will have a reference to the second `database` instance, and then `ExecutorService.getExecutor()` will return different executor instances at different times. Sometimes the `LoggingExecutor` and sometimes the `JdbcExector`.

Why was I instantiating Liquibase more than once: See [this issue](https://github.com/liquibase/liquibase/issues/2580). Instantiating a new instance worked around that caching issues (but clearly invited other issues)

**Comment by kataggart** [comment]
Thanks @leblonk for all of the detail here. We are reviewing this.

**Comment by nvoxland** [comment]
I did a simple test of

```
   @Test
    public void test() throws Exception {
        Liquibase liquibase = createLiquibase("changelogs/h2/complete/simple.changelog.xml");
        liquibase.checkLiquibaseTables(true, liquibase.getDatabaseChangeLog(), null, null);

        liquibase.setChangeLogParameter( "loginuser", testSystem.getUsername());

        StringWriter scriptWriter = new StringWriter();

        liquibase.update(this.contexts, scriptWriter);
        System.out.println(scriptWriter);

        final Statement statement = ((JdbcConnection) getDatabase().getConnection()).getUnderlyingConnection().createStatement();
        ResultSet resultSet = statement.executeQuery("select count(*) from databasechangelog");
        while (resultSet.next()) {
            System.out.println("RS: "+resultSet.getObject(1));
        }

        liquibase.update(this.contexts);

        System.out.println("AFTER:");
         resultSet = statement.executeQuery("select count(*) from databasechangelog");
        while (resultSet.next()) {
            System.out.println("RS: "+resultSet.getObject(1));
        }
    }
```
to verify you claim that it's triggered by separate Liquibase instances. That test worked correctly for me.

We are preparing for a larger refactoring to finish clearing out some of the old code styles in favor of a unified and consistent way of handling databases/caches/etc. 

If this is only something you're seeing when you're using multiple Liquibase instances, I think I'll move this to be part of that larger effort. I did find a fix for #2580 which should allow you to avoid this problem for now.

That is the correct understanding, @leblonk ?

**Comment by leblonk** [comment]
Yes, only seeing that when using multiple instances of liquibase.

**Comment by nvoxland** [comment]
Thanks. I'm going to move it into the "complicated" pile for now to handle when we do the larger focus on the HistoryService
